File: ga/4omini_result_2024_7/result/1968F_mini/48_1968F.c
use std::io::{self, BufRead};
use std::time::SystemTime;

const N: usize = 200_000;

struct RandomGenerator {
    z: u32,
}

impl RandomGenerator {
    fn new() -> Self {
        let now = SystemTime::now();
        let since_epoch = now.duration_since(std::time::UNIX_EPOCH).unwrap();
        let seed = (since_epoch.as_secs() ^ since_epoch.subsec_micros() as u64) as u32 | 1;
        
        RandomGenerator { z: seed }
    }

    fn rand(&mut self) -> u32 {
        self.z = self.z.wrapping_mul(3);
        self.z >> 1
    }
}

fn sort(ii: &mut Vec<usize>, aa: &[u32], l: usize, r: usize, rng: &mut RandomGenerator) {
    let mut l = l;
    let mut r = r;
    
    while l < r {
        let i = l;
        let mut j = l;
        let mut k = r;
        let i_ = ii[l + (rng.rand() % (r - l) as u32) as usize];
        
        while j < k {
            let c = if aa[ii[j]] != aa[i_] {
                aa[ii[j]] as i32 - aa[i_] as i32
            } else {
                ii[j] as i32 - i_ as i32
            };

            if c == 0 {
                j += 1;
            } else if c < 0 {
                ii.swap(i, j);
                j += 1;
                l += 1;
            } else {
                k -= 1;
                ii.swap(j, k);
            }
        }
        sort(ii, aa, i, l, rng);
        l = k;
    }
}

fn search(a: u32, ii: &[usize], aa: &[u32], n: usize, i: usize) -> isize {
    let mut lower = -1;
    let mut upper = n as isize + 1;

    while upper - lower > 1 {
        let h = (lower + upper) / 2;

        if aa[ii[h as usize]] < a || (aa[ii[h as usize]] == a && ii[h as usize] < i) {
            lower = h;
        } else {
            upper = h;
        }
    }
    lower
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    let mut rng = RandomGenerator::new();

    for _ in 0..t {
        let input: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace()
            .map(|x| x.parse().unwrap()).collect();
        let n = input[0];
        let q = input[1];

        let mut aa = vec![0u32; n + 1];
        for i in 1..=n {
            let value: u32 = lines.next().unwrap().unwrap().parse().unwrap();
            aa[i] = aa[i - 1] ^ value;
        }

        let mut ii: Vec<usize> = (0..=n).collect();
        sort(&mut ii, &aa, 0, n + 1, &mut rng);

        for _ in 0..q {
            let input: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace()
                .map(|x| x.parse().unwrap()).collect();
            let l = input[0] - 1;
            let r = input[1];

            let result = if aa[l] == aa[r] || ii[(search(aa[r], &ii, &aa, n, l) + 1) as usize] < 
                                                   ii[search(aa[l], &ii, &aa, n, r) as usize] {
                "YES"
            } else {
                "NO"
            };
            println!("{}", result);
        }
    }
}

File: ga/gpt3_5_result_2024_6/result/1968F_cpp_gpt/48_1968F.rs
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines().map(|line| line.unwrap());

    let t: usize = lines.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let line: Vec<usize> = lines.next().unwrap().split_whitespace().map(|x| x.parse().unwrap()).collect();
        let n = line[0];
        let q = line[1];

        let arr: Vec<usize> = lines.next().unwrap().split_whitespace().map(|x| x.parse().unwrap()).collect();

        let mut x: Vec<(usize, usize)> = vec![(0, 0); n + 1];
        x[0] = (0, 0);
        for i in 1..=n {
            x[i] = (x[i - 1].0 ^ arr[i - 1], i);
        }

        let mut x2: Vec<usize> = x.iter().map(|&(val, _)| val).collect();
        x.sort();

        for _ in 0..q {
            let query: Vec<usize> = lines.next().unwrap().split_whitespace().map(|x| x.parse().unwrap()).collect();
            let l = query[0];
            let r = query[1];

            let c = x2[r] ^ x2[l - 1];
            if c == 0 {
                println!("YES");
                continue;
            }

            let i1 = x.binary_search_by_key(&(c ^ x2[l - 1]), |&(val, _)| val).unwrap_or_else(|x| x);
            if i1 == n + 1 || x[i1].1 > r || x[i1].0 != (c ^ x2[l - 1]) {
                println!("NO");
                continue;
            }

            let i2 = x.binary_search_by_key(&x2[l - 1], |&(val, idx)| idx).unwrap_or_else(|x| x);
            if i2 == n + 1 || x[i2].1 > r || x[i2].0 != x2[l - 1] {
                println!("NO");
            } else {
                println!("YES");
            }
        }
    }
}